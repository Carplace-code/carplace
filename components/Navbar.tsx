"use client";

import { SignedIn, SignedOut, SignInButton, SignUpButton, UserButton } from "@clerk/nextjs";
import { Car, Heart, LineChart, Menu, Search } from "lucide-react";
import Link from "next/link";
import { usePathname } from "next/navigation";
import { useState } from "react";

import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";

// simple helper to join classes
const cn = (...args: (string | undefined | false)[]) => args.filter(Boolean).join(" ");

export default function Navbar() {
  const [mobileOpen, setMobileOpen] = useState(false);
  const pathname = usePathname();

  const navItems = [
    { href: "/", label: "Inicio", icon: null },
    { href: "/browse", label: "Buscar", icon: null },
    { href: "/trends", label: "Tendencias", icon: <LineChart className="h-4 w-4" /> },
  ];

  const isActive = (href: string) => (href === "/" ? pathname === "/" : pathname.startsWith(href));

  return (
    <nav className="w-full border-b bg-white shadow-sm">
      <div className="mx-auto flex max-w-7xl items-center justify-between px-4 py-3">
        {/* Logo + Search */}
        <div className="flex items-center gap-6">
          <Link href="/" className="flex items-center gap-2 text-xl font-bold">
            <Car className="h-6 w-6 text-blue-600" />
            <span>CarPlace</span>
          </Link>

          <div className="relative hidden items-center md:flex">
            <Input type="text" placeholder="Search cars, brands, or models..." className="w-80 pr-10" />
            <Search className="absolute top-2.5 right-3 h-4 w-4 text-gray-500" />
          </div>
        </div>

        {/* Desktop Nav Links */}
        <div className="hidden items-center gap-6 text-sm font-medium md:flex">
          {navItems.map(({ href, label, icon }) => (
            <Link
              key={href}
              href={href}
              className={cn(
                "flex items-center gap-1",
                isActive(href) ? "border-b-2 border-blue-600 pb-1 text-blue-600" : "text-gray-700 hover:text-blue-500",
              )}
            >
              {icon}
              {label}
            </Link>
          ))}

          <SignedOut>
            <SignInButton mode="modal">Iniciar sesión</SignInButton>
            <SignUpButton mode="modal">Crear cuenta</SignUpButton>
          </SignedOut>

          <SignedIn>
            <Link
              href="/wishlists"
              className={cn(
                "flex items-center gap-1",
                isActive("/wishlists")
                  ? "border-b-2 border-blue-600 pb-1 text-blue-600"
                  : "text-gray-700 hover:text-blue-500",
              )}
            >
              <Heart className="h-4 w-4" />
              Listas de deseos
            </Link>
            <UserButton />
          </SignedIn>
        </div>

        {/* Mobile Menu Button */}
        <div className="md:hidden">
          <Menu className="h-6 w-6 cursor-pointer" onClick={() => setMobileOpen(!mobileOpen)} />
        </div>
      </div>

      {/* Mobile Dropdown */}
      {mobileOpen && (
        <div className="space-y-3 border-t bg-white px-4 py-3 md:hidden">
          <Input type="text" placeholder="Search..." className="w-full" />

          <div className="flex flex-col gap-2 text-sm font-medium">
            {navItems.map(({ href, label, icon }) => (
              <Link
                key={href}
                href={href}
                className={cn(
                  "flex items-center gap-1 py-1",
                  isActive(href) ? "text-blue-600" : "text-gray-700 hover:text-blue-500",
                )}
              >
                {icon}
                {label}
              </Link>
            ))}

            <SignedOut>
              <Button className="w-full">Iniciar Sesión</Button>
            </SignedOut>
          </div>
        </div>
      )}
    </nav>
  );
}
