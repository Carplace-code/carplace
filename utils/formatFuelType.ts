import { FuelType } from "@prisma/client";

export default function formatFuelType(fuelType: FuelType): string {
  switch (fuelType) {
    case FuelType.gas:
      return "Gasolina";
    case FuelType.diesel:
      return "Diésel";
    case FuelType.electricity:
      return "Eléctrico";
    case FuelType.hybrid:
      return "Híbrido";
    default:
      return "Desconocido";
  }
}
