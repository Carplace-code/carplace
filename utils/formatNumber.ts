export default function formatNumber(value: number, options: { useThousandsSeparator?: boolean } = {}): string {
  if (typeof value !== "number") throw new TypeError("Value must be a number or Decimal instance");

  const { useThousandsSeparator } = options;
  const stringValue = value.toString();
  return useThousandsSeparator ? stringValue.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ".") : stringValue;
}
