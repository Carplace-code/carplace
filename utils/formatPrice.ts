export default function formatPrice(price: number | string): string {
  const localPrice = typeof price === "string" ? parseFloat(price.replace(/[^0-9.-]+/g, "")) : price;
  return Number(localPrice).toLocaleString("es-CL", {
    style: "currency",
    currency: "CLP",
  });
}
