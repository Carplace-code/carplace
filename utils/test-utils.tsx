// tests/test-utils.tsx
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { render } from "@testing-library/react";
import { ReactNode } from "react";

// Puedes extender esta función si necesitas envolver en más contextos
export function renderWithQueryClient(children: ReactNode) {
  const queryClient = new QueryClient();
  return render(<QueryClientProvider client={queryClient}>{children}</QueryClientProvider>);
}
