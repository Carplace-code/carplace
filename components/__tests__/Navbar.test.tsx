import { describe, it, vi } from "vitest";

import { renderWithQueryClient } from "@/utils/test-utils";

import Navbar from "../Navbar";

// Mock Clerk components
vi.mock("@clerk/nextjs", () => ({
  SignedIn: ({ children }: { children: React.ReactNode }) => children,
  SignedOut: ({ children }: { children: React.ReactNode }) => children,
  SignInButton: ({ children }: { children: React.ReactNode }) => <button type="button">{children}</button>,
  SignUpButton: ({ children }: { children: React.ReactNode }) => <button type="button">{children}</button>,

  UserButton: () => <div>User</div>,
}));

// Mock Next.js navigation
vi.mock("next/navigation", () => ({
  usePathname: () => "/",
  useRouter: () => ({
    push: vi.fn(),
    replace: vi.fn(),
    prefetch: vi.fn(),
    back: vi.fn(),
    forward: vi.fn(),
    refresh: vi.fn(),
  }),
}));

// Mock lucide-react icons
vi.mock("lucide-react", () => ({
  Car: () => <div>CarIcon</div>,
  Heart: () => <div>HeartIcon</div>,
  LineChart: () => <div>LineChartIcon</div>,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Menu: (props: any) => <div {...props}>MenuIcon</div>,
  Search: () => <div>SearchIcon</div>,
}));

describe("Navbar", () => {
  it("renders correctly and matches snapshot", () => {
    const { container } = renderWithQueryClient(<Navbar />);
    expect(container).toMatchSnapshot();
  });
});
