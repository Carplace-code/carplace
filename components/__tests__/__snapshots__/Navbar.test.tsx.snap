// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar > renders correctly and matches snapshot 1`] = `
<div>
  <nav
    class="w-full border-b bg-white shadow-sm"
  >
    <div
      class="mx-auto flex max-w-7xl items-center justify-between px-4 py-3"
    >
      <div
        class="flex items-center gap-6"
      >
        <a
          class="flex items-center gap-2 text-xl font-bold"
          href="/"
        >
          <div>
            CarIcon
          </div>
          <span>
            CarPlace
          </span>
        </a>
        <div
          class="relative hidden items-center md:flex"
        >
          <div
            class="container flex h-16 items-center px-4"
          >
            <div
              class="flex-1 max-w-md mx-4 relative"
            >
              <div
                class="relative"
              >
                <div>
                  SearchIcon
                </div>
                <input
                  class="file:text-foreground placeholder:text-muted-foreground selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-input flex h-9 min-w-0 rounded-md border bg-transparent px-3 py-1 text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:cursor-not-allowed focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] disabled:opacity-50 md:text-sm aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive w-80 pr-10"
                  data-slot="input"
                  placeholder="Buscar por marca, modelo o año..."
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="hidden items-center gap-6 text-sm font-medium md:flex"
      >
        <a
          class="flex items-center gap-1 border-b-2 border-blue-600 pb-1 text-blue-600"
          href="/"
        >
          Inicio
        </a>
        <a
          class="flex items-center gap-1 text-gray-700 hover:text-blue-500"
          href="/browse"
        >
          Buscar
        </a>
        <button
          type="button"
        >
          Iniciar sesión
        </button>
        <button
          type="button"
        >
          Crear cuenta
        </button>
        <a
          class="flex items-center gap-1 text-gray-700 hover:text-blue-500"
          href="/wishlists"
        >
          <div>
            HeartIcon
          </div>
          Listas de deseos
        </a>
        <div>
          User
        </div>
      </div>
      <div
        class="md:hidden"
      >
        <div
          class="h-6 w-6 cursor-pointer"
        >
          MenuIcon
        </div>
      </div>
    </div>
  </nav>
</div>
`;
